### DISCUSSION TITLE
Not all emoji tweets found with stream based on their context within the tweet

### DISCUSSION THREAD

**Discussion by bfallc** [tweepy/tweepy#1818] [Questions] [answered]
Hi, 

I've been experimenting with capturing tweets from the stream that have emoji in them. It seems that tweepy only captures tweets with emoji in certain positions relative to other text. I am using the symbols emoji (🔣) to test because it is relatively unused/unpopular on twitter. 

Here are some examples:

- Tweet (single emoji character): 🔣 
  Response: captured by tweepy in stream.

- Tweet (two or more emoji characters in row): 🔣😁🔣
  Response: **not captured by tweepy.**
  
- Tweet (two emoji separated by a space, or the target emoji separate from the other emoji): 😁 🔣 😁😁😁
  Response: captured.
 
- Tweet (mixed text and emoji, with emoji separated with a space): hello 🔣 world
  Response: captured
 
- Tweet (mixed text and emoji, with emoji touching another character or punctuation): hello world🔣 
  Response: **not captured by tweepy**

My code is below. If I think about how most emoji-bearing tweets are written, often an emoji is tacked onto the end of a tweet without a space (due to character limits on twitter), and more often than not it is in a string of multiple emoji characters, unseparated by spaces. Both those use cases would be skipped by the tweepy stream and not returned. 

Is this a flaw, or am I overlooking something? Perhaps this has to do with how tweepy parses the unicode characters as opposed to text? 

Thank you.

```
import tweepy

consumer_key = "<myconsumerkey>"
consumer_secret = "<myconsumersecret>"
access_token = "<mytoken>"
access_token_secret = "<mytokensecret>"

# Subclass Stream to print IDs of Tweets received
class IDPrinter(tweepy.Stream):

    def on_status(self, status):
        # print(status.id)

        if "qazwsx" in status.text:
            print('\n' + "qazwsx:")
            print(status.text + '\n')
         
            return
        
        if "qazedc" in status.text:
            print('\n' + "qazedc:")
            print(status.text + '\n')
   
            return

        if "\U0001F523" in status.text:
            print('\n' + "symbols emoji (unicode):")
            print(status.text + '\n')
   
            return

        # also works:
        # if "🔣" in status.text:
        #     print('\n' + "symbols emoji (emoji):")
        #     print(status.text + '\n')
        #     return

# Initialize instance of the subclass
printer = IDPrinter(
  consumer_key, consumer_secret,
  access_token, access_token_secret
)

# Filter realtime Tweets by keyword
printer.filter(track=['👋', 'qazwsx', 'qazedc', '💘', '🔣', '\U0001F523'])
```

**Comment by Harmon758** [comment, marked as answer]
This is how Twitter's API works and is not an issue with Tweepy.

The "text" that you are referring to is also composed of Unicode characters. Any character that you're able to input as a Tweet is going to be a Unicode character, including the letters a-z, etc. Accordingly, emoji are handled in the same way as any other character.

Streaming Tweets by filtering by keyword returns Tweets that match the keyword as a word, not as part of another word. This means that the keyword needs to be space-separated from other characters. In the same way that filtering by the keyword "hi" will not return all Tweets with any word that contains "hi" in them, e.g. "history", "chime", "sophisticated", etc., filtering by a specific emoji as a keyword will not return Tweets where the only instances of the emoji are those where it's part of another "word", whether that be composed of only emoji or a combination of emoji and "text".

(In the documentation for streaming with Twitter API v2, this is referred to as a "tokenized match":
> Matches a keyword within the body of a Tweet. This is a tokenized match, meaning that your keyword string will be matched against the tokenized text of the Tweet body. Tokenization splits words based on punctuation, symbols, and Unicode basic plane separator characters.
> 
> For example, a Tweet with the text “I like coca-cola” would be split into the following tokens: I, like, coca, cola. These tokens would then be compared to the keyword string used in your rule.

> Matches an emoji within the body of a Tweet. Similar to a keyword, emojis are a tokenized match, meaning that your emoji will be matched against the tokenized text of the Tweet body.

https://developer.twitter.com/en/docs/twitter-api/tweets/filtered-stream/integrate/build-a-rule)

**Reply by bfallc** [reply]
Thank you for this explanation, makes sense.
